internal/pkg/policy/policy_output.go

```diff
@@ -137,6 +192,104 @@ func (p *PolicyOutput) Prepare(ctx context.Context, zlog zerolog.Logger, bulker
 	return nil
 }
 
+func fetchAPIKeyRoles(ctx context.Context, b bulk.Bulk, apiKeyID string) (*RoleT, error) {
+	res, err := b.APIKeyRead(ctx, apiKeyID, true)
+	if err != nil {
+		return nil, err
+	}
+
+	roleMap, err := smap.Parse(res.RoleDescriptors)
+	if err != nil {
+		return nil, err
+	}
+	r := &RoleT{
+		Raw: res.RoleDescriptors,
+	}
+
+	// Stable hash on permissions payload
+	if r.Sha2, err = roleMap.Hash(); err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func mergeRoles(zlog zerolog.Logger, old, new *RoleT) (*RoleT, error) {
+	if old == nil {
+		return new, nil
+	}
+	if new == nil {
+		return old, nil
+	}
+
+	oldMap, err := smap.Parse(old.Raw)
+	if err != nil {
+		return nil, err
+	}
+	if oldMap == nil {
+		return new, nil
+	}
+
+	newMap, err := smap.Parse(new.Raw)
+	if err != nil {
+		return nil, err
+	}
+
+	if newMap == nil {
+		return old, nil
+	}
```

Comment:
same as above

[Consistency]
```suggestion
	if newMap == nil {
		return old, nil
	}
```